source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1461 @@
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ directive.
+
+Test runners may choose to implement the entity map in a fashion most suited to
+their language, but implementations MUST enforce both uniqueness of entity names
+and referential integrity when fetching an entity. Test runners MUST raise an
+error if an attempt is made to store an entity with a name that already exists
+in the map and MUST raise an error if an entity is not found for a name or is
+found but has an unexpected type.
+
+Consider the following examples::
+
+    # Error due to a duplicate name (client0 was already defined)
+    createEntities:
+      - client: { id: client0 }
+      - client: { id: client0 }
+
+    # Error due to a missing entity (client1 is undefined)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client1 }
+
+    # Error due to an unexpected entity type (session instead of client)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client0 }
+      - session: { id: session1, client: session0 }
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as a document by the test runner. This section defines the top-level
+keys for that document and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies. Test runners will use this to determine compatibility
+  (i.e. whether and how the test file will be interpreted). The format of this
+  string is defined in `Version String`_.
+
+.. _runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. These requirements
+  may be overridden on a per-test basis by `test.runOn <test_runOn_>`_. Test
+  runners MUST skip a test if its requirements are not met.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `runOnRequirement`_.
+
+.. _createEntities:
+
+- ``createEntities``: Optional array of documents. List of entities (e.g.
+  client, collection, session objects) that should be created before each test
+  case is executed. The structure of each document is defined in `entity`_.
+
+.. _collectionName:
+
+- ``collectionName``: Optional string. Name of collection under test. This is
```

Comment:
`collectionData`, which is used for both `initialData` and `test.outcome`, purposely solicits name strings instead of entities because it operates using the internal MongoClient. When it comes to writing tests for this spec (not to be confused with spec tests that will _use_ the format), I did plan to create a test file with no entities and allow the test runner to verify that the documents created by `initialData` would be found for `test.outcome` (no operations would actually need to operate on a collection under test).

That said, there's no reason the entities and `collectionData` cannot utilize anchors/aliases to avoid copy-pasting string literals. The only consideration is that anchors must be defined before aliases, so whichever of `initialData` or `createEntities` appeared first in the test file would be able to define the anchor.

I was fond of the original behavior where a test runner could generate its own names as a fallback, since there are a number of tests where the names are irrelevant (not used in assertions). In the case of PHP, we try to run our test suite in the same database ("test") and this made it easier to do so for tests that didn't need to specify a custom database (it's more common to customize the collection name). However, if all tests are going to be ported over to this format, that would also give us a chance to standardize on the names and the spec could even provide guidance on names to use (thinking more like "e.g." instead of "SHOULD").

I'll admit that I probably have a bias from the MMAPv1 days, where using a variety of databases had a direct impact on the file system (I also remember drops being slower). I don't think this is relevant for WiredTiger, as my local server's data directory has a bunch of `collection-#` and `index-#` files in it and is no longer organized by database.

I'll leave this thread open so we can discuss in the meeting tomorrow. I suppose the only remaining selling point of preserving the generation fallback behavior would be to have more concise YAML/JSON.

Reply:
We decided to drop the top-level fields and test runner generation logic (when omitted) and require explicit names for both entities and `collectionData` structure (for `initialData` and `outcome`). YAML anchors can be used to reduce copypasta.